utils/pattern/src/pattern/mod.rs

```diff
@@ -0,0 +1,185 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+mod error;
+
+use crate::parser::error::ParserError;
+use crate::token::PatternToken;
+use crate::{
+    interpolator::{InterpolatedKind, Interpolator, InterpolatorError},
+    parser::{Parser, ParserOptions},
+    replacement::ReplacementProvider,
+};
+pub use error::PatternError;
+use std::{
+    convert::{TryFrom, TryInto},
+    fmt::{Debug, Display, Write},
+    str::FromStr,
+};
+
+/// `Pattern` stores the result of parsing operation as a vector
+/// of [`PatternToken`] elements.
+///
+/// # Type parameters
+///
+/// - `P`: The type of the placeholder used as a key for the [`ReplacementProvider`].
+///
+/// # Lifetimes
+///
+/// - `p`: The life time of an input string slice to be parsed.
+///
+/// [`ReplacementProvider`]: crate::ReplacementProvider
+pub struct Pattern<'s, P>(pub(crate) Vec<PatternToken<'s, P>>);
+
+impl<'s, P> Pattern<'s, P> {
+    /// Interpolates the `Pattern` with provided replacements and returns
+    /// a [`InterpolatedPattern`] structure.
+    ///
+    /// For allocation-free interpolation, see `interpolate_to_string` or
+    /// `interpolate_to_write`.
+    ///
+    /// For lower level interpolation iterator see [`Interpolator`].
+    pub fn interpolate<'i, E, R>(
+        &'i self,
+        replacements: &'i R,
+    ) -> Result<InterpolatedPattern<'i, 's, E>, PatternError<R::Key>>
+    where
+        R: ReplacementProvider<'i, E, Key = P>,
+        P: Debug + FromStr + Clone,
+        <P as FromStr>::Err: Debug,
+    {
+        Interpolator::new(self, replacements)
+            .try_into()
+            .map_err(Into::into)
+    }
+
+    /// Interpolates the `Pattern` with provided replacements and a new
+    /// [`String`].
+    ///
+    /// For buffer write interpolation, see `interpolate_to_write`.
+    ///
+    /// For lower level interpolation iterator see [`Interpolator`].
+    pub fn interpolate_to_string<'i, E, R>(
+        &'i self,
+        replacements: &'i R,
+    ) -> Result<String, PatternError<R::Key>>
+    where
+        R: ReplacementProvider<'i, E, Key = P>,
+        P: Debug + FromStr + Clone,
+        <P as FromStr>::Err: Debug,
+        E: 'i + Display,
+    {
+        let mut result = String::new();
+        self.interpolate_to_write(replacements, &mut result)?;
+        Ok(result)
+    }
+
+    /// Interpolates the `Pattern` writing the result into a buffer.
+    pub fn interpolate_to_write<'i, E, R, W>(
+        &'i self,
+        replacements: &'i R,
+        f: &mut W,
+    ) -> Result<(), PatternError<R::Key>>
+    where
+        R: ReplacementProvider<'i, E, Key = P>,
+        P: Debug + FromStr + Clone,
+        <P as FromStr>::Err: Debug,
+        E: 'i + Display,
```

Comment:
I'm okay adding support for `Writeable` in this PR, but I'd prefer not to have to figure out how to handle all blank implementations should work, where should they live and so on in this PR.

Reply:
Okay... well, eventually this should be backed by FormattedStringBuilder (#135) anyhow, so that we get formatToParts support.  I'm happy that you added Writeable elsewhere where you could though.